src/hooks/carousel/__tests__/useCarousel.browser.spec.js

```diff
@@ -0,0 +1,119 @@
+import { Box } from '&test/browser/utils/Box';
+import { mountHookWithUI } from '&test/browser/utils/mountHookWithUI';
+import { useCarousel } from '../useCarousel';
+
+const mainStyle = `
+  .list {
+    width: 700px;
+    max-width: 100%;
+  }
+`;
+
+const useWrappedHook = (result) => {
+  const ref = useRef();
+  const [selected, setSelected] = useState(0);
+  result.current = { selected, setSelected };
+
+  const carousel = useCarousel(ref, { selected, onSelect: setSelected, numTiles: 4 }, []);
+
+  return (
+    <>
+      <div>
+        <button type="button" {...carousel.props.arrow.left}>left</button>
+        <button type="button" {...carousel.props.arrow.right}>right</button>
+      </div>
+      <div className="list" ref={ref} {...carousel.props.list}>
+        <Box id="box-1" {...carousel.props.tile(0)}><button type="button">yolo</button></Box>
+        <Box id="box-2" {...carousel.props.tile(1)}><button type="button">yolo</button></Box>
+        <Box id="box-3" {...carousel.props.tile(2)}><button type="button">yolo</button></Box>
+        <Box id="box-4" {...carousel.props.tile(3)}><button type="button">yolo</button></Box>
+        <div className="margin-fix" data-lwc-ignore />
+      </div>
+      <p>
+        {carousel.pages.current} / {carousel.pages.total}
+      </p>
+    </>
+  );
+};
+
+const mountTest = (props) =>
+  mountHookWithUI(useWrappedHook, props, mainStyle);
+
+it('selects tile on click or with arrow keys', () => {
+  mountTest().its('result').as('result');
+
+  cy.get('@result').its('current.selected').should('toEqual', 0);
+  cy.get('#box-1').should('toHaveClass', 'lwc-selected');
+  cy.get('#box-1').siblings().should('not.toHaveClass', 'lwc-selected');
+
+  cy.get('#box-2 button').click();
+  cy.get('@result').its('current.selected').should('toEqual', 1);
+  cy.get('#box-2').should('toHaveClass', 'lwc-selected');
+  cy.get('#box-2').siblings().should('not.toHaveClass', 'lwc-selected');
+
+  cy.get('#box-3 button').click();
+  cy.get('@result').its('current.selected').should('toEqual', 2);
+  cy.get('#box-3').should('toHaveClass', 'lwc-selected');
+  cy.get('#box-3').siblings().should('not.toHaveClass', 'lwc-selected');
+
+  cy.get('#box-3 button').type('{rightarrow}');
+  cy.get('@result').its('current.selected').should('toEqual', 3);
+  cy.get('#box-4').should('toHaveClass', 'lwc-selected');
+  cy.get('#box-4').siblings().should('not.toHaveClass', 'lwc-selected');
+
+  cy.get('#box-4 button').type('{leftarrow}');
+  cy.get('@result').its('current.selected').should('toEqual', 2);
+  cy.get('#box-3').should('toHaveClass', 'lwc-selected');
+  cy.get('#box-3').siblings().should('not.toHaveClass', 'lwc-selected');
+});
+
+it('calculates the current and total pages', () => {
+  cy.viewport(800, 500);
+  mountTest();
+
+  cy.findByText('1 / 2').should('exist');
+
+  cy.get('.list').invoke('scrollLeft', 500);
+  cy.findByText('2 / 2').should('exist');
+
+  cy.viewport(500, 500);
+  cy.findByText('3 / 4').should('exist');
```

Comment:
Is the viewport an "instantaneous" change? Or does the find by wait until the viewport has been updated?

Reply:
Yes and no 😄 , all cypress commands are asynchronous behind the scenes. They actually only enqueue the action & then cypress runs them all after the `test()` block is run. So the viewport is completely changed once the `viewport` command completes. Also, most cypress commands will auto-retry if the command _or the following assertion fails_.

https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Asynchronous
https://docs.cypress.io/guides/core-concepts/retry-ability.html#Commands-vs-assertions